pyteal/compiler/compiler.py

```diff
@@ -133,26 +133,64 @@ def compileSubroutine(
             ast = Return(ast)
 
     options.setSubroutine(currentSubroutine)
+
     start, end = ast.__teal__(options)
     start.addIncoming()
     start.validateTree()
 
-    start = TealBlock.NormalizeBlocks(start)
-    start.validateTree()
+    if (
+        currentSubroutine is not None
+        and currentSubroutine.get_declaration().deferred_expr is not None
+    ):
+        # this represents code that should be inserted before each retsub op
+        deferred_expr = cast(Expr, currentSubroutine.get_declaration().deferred_expr)
+
+        # creating a function so that we can have a new copy of the blocks every time we want to
+        # insert them -- you cannot have the same block in two different places in the control flow
+        # graph
+        def get_deferred_blocks():
+            return deferred_expr.__teal__(options)
+
+        for block in TealBlock.Iterate(start):
+            if not any(op.getOp() == Op.retsub for op in block.ops):
+                continue
+
+            if len(block.ops) != 1:
+                # we expect all retsub ops to be in their own block at this point since
+                # TealBlock.NormalizeBlocks has not yet been used
+                raise TealInternalError(
+                    f"Expected retsub to be the only op in the block, but there are {len(block.ops)} ops"
+                )
 
-    order = sortBlocks(start, end)
-    teal = flattenBlocks(order)
+            deferred_start, deferred_end = get_deferred_blocks()
```

Comment:
nit: is it possible to move line 152 `__teal__` to this line?

Reply:
Good idea, the function approach made sense initially but I think we can get rid of it now